pkg/fleet/installer/service/apm_inject.go

```diff
@@ -37,110 +34,126 @@ dogstatsd_socket: %s
 	var err error
 	span, ctx := tracer.StartSpanFromContext(ctx, "setup_injector")
 	defer span.Finish(tracer.WithError(err))
-	installer := &apmInjectorInstaller{
-		installPath: "/opt/datadog-packages/datadog-apm-inject/stable",
-	}
+	installer := newAPMInjectorInstaller(injectorPath)
 	return installer.Setup(ctx)
 }
 
 // RemoveAPMInjector removes the APM injector
 func RemoveAPMInjector(ctx context.Context) {
 	span, ctx := tracer.StartSpanFromContext(ctx, "remove_injector")
 	defer span.Finish()
-	installer := &apmInjectorInstaller{
-		installPath: "/opt/datadog-packages/datadog-apm-inject/stable",
-	}
+	installer := newAPMInjectorInstaller(injectorPath)
 	installer.Remove(ctx)
 }
 
+func newAPMInjectorInstaller(path string) *apmInjectorInstaller {
+	a := &apmInjectorInstaller{
+		installPath: path,
+	}
+	a.ldPreloadFileInstrument = newFileMutator(ldSoPreloadPath, a.setLDPreloadConfigContent, nil, nil)
+	a.ldPreloadFileUninstrument = newFileMutator(ldSoPreloadPath, a.deleteLDPreloadConfigContent, nil, nil)
+	a.agentConfigSockets = newFileMutator(datadogConfigPath, a.setAgentConfigContent, nil, nil)
+	a.dockerConfigInstrument = newFileMutator(dockerDaemonPath, a.setDockerConfigContent, nil, nil)
+	a.dockerConfigUninstrument = newFileMutator(dockerDaemonPath, a.deleteDockerConfigContent, nil, nil)
+	return a
+}
+
 type apmInjectorInstaller struct {
-	installPath string
+	installPath               string
+	ldPreloadFileInstrument   *fileMutator
+	ldPreloadFileUninstrument *fileMutator
+	agentConfigSockets        *fileMutator
+	dockerConfigInstrument    *fileMutator
+	dockerConfigUninstrument  *fileMutator
 }
 
 // Setup sets up the APM injector
 func (a *apmInjectorInstaller) Setup(ctx context.Context) error {
 	var err error
+	var rollbackAgentConfig, rollbackLDPreload, rollbackDockerConfig func() error
 	defer func() {
 		if err != nil {
-			a.Remove(ctx)
+			// todo propagate rollbacks until success of package installation
+			if rollbackLDPreload != nil {
+				if err := rollbackLDPreload(); err != nil {
+					log.Warnf("Failed to rollback ld preload: %v", err)
+				}
+			}
+			if rollbackAgentConfig != nil {
+				if err := rollbackAgentConfig(); err != nil {
+					log.Warnf("Failed to rollback agent config: %v", err)
+				}
+			}
+			if rollbackDockerConfig != nil {
+				if err := rollbackDockerConfig(); err != nil {
+					log.Warnf("Failed to rollback docker config: %v", err)
+				}
+			}
 		}
 	}()
-	if err := a.setAgentConfig(ctx); err != nil {
-		return err
-	}
-	if err := a.setRunPermissions(); err != nil {
-		return err
-	}
-	if err := a.setLDPreloadConfig(ctx); err != nil {
+
+	rollbackAgentConfig, err = a.setupSockets(ctx)
+	if err != nil {
 		return err
 	}
-	if err := a.setDockerConfig(ctx); err != nil {
+
+	rollbackLDPreload, err = a.ldPreloadFileInstrument.mutate()
+	if err != nil {
 		return err
 	}
-	return nil
+
+	// TODO only instrument docker if DD_APM_INSTRUMENTATION_ENABLED=docker
+	// is set
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ Question:** ‏Why? This would block remote installation of the docker instrumentation later on

Reply:
customers currently configure DD_APM_INSTRUMENTATION_ENABLED=(all|host|docker), we should be first compatible with it, and then make remote_installation work as all